### DISCUSSION TITLE
Different results obtained with different functions for scoring NER

### DISCUSSION THREAD

**Discussion by michelemarzollo** [explosion/spaCy#8585] [Help: Coding & Implementations] [answered]
Hello,

I am trying to evaluate the performance of some models for NER.
I realized that using the CLI `evaluate` command I obtain different results from those obtained using the Scorer from code. I also implemented the scoring myself, and the result is different from both of them.

I would like to know if I am doing something wrong, or if there is a reason for which this behavior is correct.

I created a few functions that evaluate the same model in four different ways:

- I compute the score counting TP, FN, FP.
- I create examples and use the spacy Scorer
- I use Spacy evaluate function from code
- I use the CLI command

Here is the code (I added the alignment, but it does not have any effect):
```python
import os

import spacy
from spacy.tokens import DocBin
from spacy.training import Example, Alignment
from spacy.scorer import Scorer

model = "trained_models/large_ner_scratch_conll/model-best"
test_data = "datasets/spacy/conll2003/test.spacy"


def manual_computation():
    tp = 0
    fp = 0
    fn = 0

    nlp = spacy.load(model)
    doc_bin = DocBin().from_disk(test_data)
    docs = doc_bin.get_docs(nlp.vocab)
    for idx, gold_doc in enumerate(docs):
        truth = [(ent.start_char, ent.end_char, ent.label_) for ent in gold_doc.ents]

        pred_doc = nlp(gold_doc.text)
        pred_ents = [(ent.start_char, ent.end_char, ent.label_) for ent in pred_doc.ents]

        pred_ents = set(pred_ents)
        truth = set(truth)

        false_positives = pred_ents - truth
        false_negatives = truth - pred_ents

        true_positives = truth.intersection(pred_ents)

        tp += len(true_positives)
        fn += len(false_negatives)
        fp += len(false_positives)

    precision = tp / (tp + fp)
    recall = tp / (tp + fn)
    f_score = 2 * precision * recall / (precision + recall)
    print("Precision:\t{:.4f}".format(precision))
    print("Recall:\t\t{:.4f}".format(recall))
    print("F-score:\t{:.4f}".format(f_score))


def use_scorer():
    examples = []
    nlp = spacy.load(model)
    doc_bin = DocBin().from_disk(test_data)
    docs = doc_bin.get_docs(nlp.vocab)
    for idx, gold_doc in enumerate(docs):
        text = gold_doc.text
        pred_doc = nlp(text)

        gold_tokens = [tok.text for tok in gold_doc]
        pred_tokens = [tok.text for tok in pred_doc]
        alignment = Alignment.from_strings(pred_tokens, gold_tokens)

        examples.append(Example(pred_doc, gold_doc, alignment=alignment))
    scorer = Scorer(nlp)
    results = scorer.score(examples)
    print("Precision:\t{:.4f}".format(results['ents_p']))
    print("Recall:\t\t{:.4f}".format(results['ents_r']))
    print("F-score:\t{:.4f}".format(results['ents_f']))


def use_evaluate():
    examples = []
    nlp = spacy.load(model)
    doc_bin = DocBin().from_disk(test_data)
    docs = doc_bin.get_docs(nlp.vocab)
    for idx, gold_doc in enumerate(docs):
        text = gold_doc.text
        pred_doc = nlp(text)

        gold_tokens = [tok.text for tok in gold_doc]
        pred_tokens = [tok.text for tok in pred_doc]
        alignment = Alignment.from_strings(pred_tokens, gold_tokens)

        examples.append(Example(pred_doc, gold_doc, alignment=alignment))
    results = nlp.evaluate(examples)
    print("Precision:\t{:.4f}".format(results['ents_p']))
    print("Recall:\t\t{:.4f}".format(results['ents_r']))
    print("F-score:\t{:.4f}".format(results['ents_f']))


def use_cli():
    os.system(f"spacy evaluate {model} {test_data}")


if __name__ == '__main__':
    print("Results from manual computation:\n")
    manual_computation()
    print("\nResults from spacy Scorer:\n")
    use_scorer()
    print("\nResults from spacy evaluate method:\n")
    use_evaluate()
    print("\nResults from spacy evaluate recipe (CLI):\n")
    use_cli()

```
The output obtained is the following:

```
Results from manual computation:

Precision:	0.8390
Recall:		0.8339
F-score:	0.8364

Results from spacy Scorer:

Precision:	0.8489
Recall:		0.8438
F-score:	0.8464

Results from spacy evaluate method:

Precision:	0.8489
Recall:		0.8438
F-score:	0.8464

Results from spacy evaluate recipe (CLI):

ℹ Using CPU

================================== Results ==================================

TOK     -    
NER P   86.96
NER R   85.39
NER F   86.17
SPEED   26116


=============================== NER (per type) ===============================

           P       R       F
LOC    89.76   91.97   90.85
PER    89.72   88.56   89.14
MISC   83.39   76.50   79.79
ORG    82.65   79.47   81.03

```

The only results that are the same are the second and the third ones.

Moreover, I don't understand how the `Language.evaluate` method works: are the predictions set in the Examples evaluated, or does the model compute again the entities? After changing either the model or the predictions in the Examples, I get always different results.

Thank you for being always helpful and fast with your answers!

**Comment by adrianeboyd** [comment, marked as answer]
Hi, my first guess would be that you're seeing side effects of unintentionally applying the `ner` component twice to the same docs. If you apply the same `ner` model twice, it might add additional entity spans where the current doc has `O`. It won't modify any existing spans, but to allow combinations like `entity_ruler` + `ner`, the `ner` component is allowed to overwrite `O`.

If you're calling the `score` method directly with examples, then you want to construct examples where the `predicted` doc contains the exact annotations you want to evaluate. `score` doesn't apply the pipeline, it just compares `predicted` vs. `reference` using the scoring methods provided by the components in the pipeline.

If you're calling `nlp.evaluate`, then you want the `predicted` docs to be as blank as your normal pipeline input, so typically created with `nlp.make_doc(text)`. `nlp.evaluate` applies the pipeline (to an internal copy of the predicted docs so they're not modified in the original `examples` list) and then calls the scorer.

(There are cases where you want to provide some layers of gold annotation in the `predicted` doc to particular components, so this is why it's not necessarily always just plain `nlp.make_doc`.)

One remaining issue might be how your reference `DocBin` was created. There was a bug in `Example.from_dict` that led to docs without any entities having "missing" rather than `O` annotation, and the built-in scorer treats `O` differently from missing annotation.

You can see the full scoring method here:

https://github.com/explosion/spaCy/blob/3b1d5350d0016ebcd054b834d55e9ad857ebfe46/spacy/scorer.py#L714-L756

**Reply by michelemarzollo** [reply]
Thank you very much for the answer.

There are two things that are still not clear to me:
- When am I applying the same ner model twice? As far as I understand it could happen only when i create the examples and then I use the `evaluate` method. But the result there is the same as the one obtained by the scorer, that does not recompute the entities.
- The DocBin was created starting from the CoNLL dataset and using the `spacy convert` command. Unless the bug was not already corrected in the latest version, I still cannot understand why the result obtained via command line is different than the one obtained with the with the `Scorer.score` method.

Anyways, thank you again for the clarification on how the methods work.

**Reply by adrianeboyd** [reply]
I don't think the `Example.from_dict` bug is an issue for `spacy convert` conversions because they create the `Doc`s in other ways, but it could be an issue if you're using it in a custom conversion scripts.

I think this is where you're applying it twice:

```python
def use_evaluate():
    examples = []
    nlp = spacy.load(model)
    doc_bin = DocBin().from_disk(test_data)
    docs = doc_bin.get_docs(nlp.vocab)
    for idx, gold_doc in enumerate(docs):
        text = gold_doc.text
        pred_doc = nlp(text) # <---- this is assigning entities in the predicted doc

        gold_tokens = [tok.text for tok in gold_doc]
        pred_tokens = [tok.text for tok in pred_doc]
        alignment = Alignment.from_strings(pred_tokens, gold_tokens)

        examples.append(Example(pred_doc, gold_doc, alignment=alignment))
    results = nlp.evaluate(examples) # <---- this is running "ner" on the same doc again
```

You want to use `nlp.make_doc` instead when constructing the predicted doc. You can leave out the steps with the alignment, it's not going to make a difference.

I tried this with our eval dataset and `en_core_web_sm` and got the same results for everything except the first manual case, which is related to how it handles missing annotation in the reference docs. You can check your data with something like:

```python
import spacy
from spacy.tokens import DocBin

nlp = spacy.blank("en")
docs = DocBin().from_disk("file.spacy").get_docs(nlp.vocab)
for doc in docs:
    assert doc.has_annotation("ENT_IOB", require_complete=True)
```

With the OntoNotes eval data, it looks like the `ner` model doesn't add any entities in the second pass, but I think that can depend a lot on the model and data. I know that it's allowed by the `ner` component, but maybe that was a red herring in terms of your example here.

I can look into this further if you can provide a model + data (or example with a pretrained pipeline like `en_core_web_sm`) that shows a discrepancy between `nlp.evaluate` and `spacy evaluate`, but currently I can't replicate it.

**Reply by michelemarzollo** [reply]
Thank you for the explanation. Things are more clear to me now.

It's surprising to me that you did not obtain discrepancies with the OntoNotes dataset. I do not have that dataset, but I always obtained different results with other datasets.

In order to provide you with an example with a pretrained pipeline, I will use `fr_core_news_sm` and a small subsection of the wikiner wp3 dataset, already converted into spacy format with the `spacy convert` command (you can find it here attached [test_small.zip](https://github.com/explosion/spaCy/files/6803147/test_small.zip)).

I used `nlp.make_doc` as you suggested, but there is no difference in the result, so I will not post again the code. You can just change lines 8 and 9 with:

```python
model = "fr_core_news_sm"
test_data = "test_small.spacy"
```
The result I obtain is the following (I excluded the manual computation part):
```
Results from spacy Scorer:

Precision:	0.8675
Recall:		0.8821
F-score:	0.8748

Results from spacy evaluate method:

Precision:	0.8675
Recall:		0.8821
F-score:	0.8748

Results from spacy evaluate recipe (CLI):

ℹ Using CPU

================================== Results ==================================

TOK      -    
TAG      -    
POS      -    
MORPH    -    
LEMMA    -    
UAS      -    
LAS      -    
NER P    91.08
NER R    90.10
NER F    90.59
SENT P   -    
SENT R   -    
SENT F   -    
SPEED    6494 


=============================== NER (per type) ===============================

           P       R       F
LOC    87.21   89.99   88.58
MISC   87.40   82.13   84.68
PER    96.89   94.06   95.45
ORG    83.62   84.35   83.98

```
I hope this can help. Thank you very much again for your commitment.

**Reply by adrianeboyd** [reply]
Thanks for the example, now I can see what's going on!

`spacy evaluate` loads the data using `Corpus` rather than only from `Example(nlp.make_doc(reference.text), reference)`, which results in different tokenizations for the predicted docs. The tokenization differences lead to different predictions, so the evaluation results are different.

Because a lot of NER data doesn't include information about whether there was whitespace between tokens (often one token per line), the converted docs end up with a setting enabled called `doc.has_unknown_spaces`, so the training and evaluation steps can keep track of whether it makes sense to use `doc.text` as a "real" text or not. When `Corpus` sees this flag on the doc loaded from the `DocBin`, `example.predicted` is constructed using the gold tokenization from the reference doc rather than the tokenization from the pipeline.

This can make a pretty big difference in French because the underlying `doc.text` from this conversion looks like this:

```none
L' âge hyborien se situe entre la chute de l' Atlantide et l' essor des anciennes civilisations que l' on connaît ( Sumer , l' Égypte antique , la Grèce antique , etc .
```

The texts looks like space-separated tokens, but only `doc.has_unknown_spaces` tells you whether this is real whitespace or not. If you tokenize this again with `nlp.make_doc` with the default French tokenizer, you end up with a lot of apostrophes split off from `l` and `d` as separate tokens, which isn't something the model saw as separate tokens while training, so the predictions are worse.

If you're curious, here's exactly what `Corpus` does:

https://github.com/explosion/spaCy/blob/8233359225df4b959a9d856ef208a66756f73502/spacy/training/corpus.py#L143-L156

The lack of `TOK` evaluation is a clue that I missed when looking at this. `has_unknown_spaces` is used to decide whether you can evaluate the tokenization accuracy based on a doc, so if none of the docs have realistic raw texts, the evaluation results in `-` rather than a float.

The differences for English are much, much smaller than for French when you tokenize space-separated tokens again with the default tokenizer, so the evaluation results are pretty much identical for both methods.

**Reply by michelemarzollo** [reply]
Thank you very much, now everything is clear! Just as a side note, I obtain significant differences also with English and datasets other than WikiNer (as you can see in the original question). I guess the difference is that OntoNotes also provides plain text.

**Reply by adrianeboyd** [reply]
I can believe that there are English datasets where you'd see a difference in the scoring after retokenizing, but we do have texts for most (although not all) of the OntoNotes data.

Now to figure out where to document this better, since it's kind of confusing and non-obvious...
